### PR TITLE
Add codecov coverage report

### LINKED ISSUE #191: Enable test coverage
**Description**

To improve the testing health of the project, we will add a [codecov](https://about.codecov.io/) integration s.t. all PRs with passing tests will recieve a codecov report. We will also fail any PR where code coverage goes down more than 2%.

**Acceptance criteria**
* Users should have the ability to locally run a code coverage plugin to get reports without submitting to CI
* CI should automatically generate a code coverage report when users submit a PR with passing tests
* As part of the pull request, evaluate the coverage and fail if the coverage goes down by more than -2%
* Enable [`code-cov`](https://about.codecov.io/) on the Github repo

### PR DESCRIPTION
**Pull request by dimberman**
Description

To improve the testing health of the project, we will add a [codecov](https://about.codecov.io/) integration s.t. all PRs with passing tests will recieve a codecov report. We will also fail any PR where code coverage goes down more than 2%.

Acceptance criteria

- [x]  Users should have the ability to locally run a code coverage plugin to get reports without submitting to CI
- [x] CI should automatically generate a code coverage report when users submit a PR with passing tests
- [x] As part of the pull request, evaluate the coverage and fail if the coverage goes down by more than -2%
- [x] Enable [code-cov](https://about.codecov.io/) on the Github repo

Resolves  #191

### CODE REVIEW

**Comment by dimberman** [comment]
@tatiana it might not be so simple for us to set up the rule around "fail test if coverage goes down", since we get our coverage information piecemeal from each test. Can we set a general rule among reviewers that we don't merge a significant reduction of code coverage without a good reason? I worry that figuring out how to get this to work considering the multi-step process might be a rabbit hole.

**Comment by tatiana** [comment]
> @tatiana it might not be so simple for us to set up the rule around "fail test if coverage goes down", since we get our coverage information piecemeal from each test. Can we set a general rule among reviewers that we don't merge a significant reduction of code coverage without a good reason? I worry that figuring out how to get this to work considering the multi-step process might be a rabbit hole.

@dimberman have you tried using `codecov.yml` and the threshold option? https://docs.codecov.com/docs/common-recipe-list

**Comment by utkarsharma2** [comment]
@dimberman can we have a test case which can call a db/table where the previous coverage value is stored, and compare it with current value? 

And when a PR is successfully merged we can update the value in db via GitHub actions.

**Comment by dimberman** [comment]
@tatiana I can add a codecov.yml but I'm not certain if that will cause failures as individual test-workers will show notably lower test coverage before they are all merged. For now I'll add it and if we see failures due to code coverage we'll just go from there.

File: .github/workflows/ci.yaml
Comment (tatiana) [inline .github/workflows/ci.yaml:103]: Please, add `--cov-branch` as well (if possible both in the CI and in the dev documentation)
https://pytest-cov.readthedocs.io/en/latest/config.html
https://coverage.readthedocs.io/en/6.3.1/branch.html

**Comment by dimberman** [comment]
@tatiana addressed your points

File: pyproject.toml
Comment (tatiana) [inline pyproject.toml:26]: @dimberman why `coverage` isn't within the `tests` dependencies?

File: docs/DEVELOPMENT.md
Comment (tatiana) [inline docs/DEVELOPMENT.md:90]: Probably worth adding  `--cov-branch` here as well, to highlight in case there are implicit `else` statements within the package/module of interest

**Review by tatiana — APPROVED** [review]
@dimberman amazing work!
Two last minor comments inline, and one question:

When I tried to see the full report, by clicking the link [Continue to review full report at Codecov](https://codecov.io/gh/astro-projects/astro/pull/203?src=pr&el=continue&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=astro-projects). `codecov.io` gave this error:
Screenshot from 2022-03-13 10-41-27

This happened after authoring the app on the `astro-projects` repos. Do you have any idea of what the issue may be?

**Comment by dimberman** [comment]
@tatiana I'm hoping that the fix there is that we need to merge the first PR to set things up. If the problem persists we will make a ticket of it

**Pull request merged**
